### PR TITLE
Added inspector for HyperlinkedRelatedField

### PR DESCRIPTION
**Pull request by stevenbal**
required for https://github.com/VNG-Realisatie/verzoeken-api/pull/4
related issue: https://github.com/VNG-Realisatie/gemma-zaken/issues/1405#issuecomment-639543611

@sergei-maertens `in_te_trekken_verzoek` was missing the `min_length` and `max_length` ([1 .. 1000]) in the API spec, I'm not sure if this is where I should add it though.

### CODE REVIEW

**Comment by sergei-maertens** [comment]
Can't you specify that on the serializer with `extra_kwargs` rather than needing a custom inspector for it?

**Comment by stevenbal** [comment]
> Can't you specify that on the serializer with `extra_kwargs` rather than needing a custom inspector for it?

for some reason, if I specify `min_length` or `max_length` via `extra_kwargs`, drf_yasg doesn't seem to pick it up, instead it actually fails to create the schema for the component altogether

**Comment by sergei-maertens** [comment]
@stevenbal yeah - I think the implementation in Open Zaak should land in vng-api-common instead :)

**Comment by joeribekker** [comment]
@sergei-maertens 

Can you describe what @stevenbal needs to do here. You were still doubting about empty string vs null for making URLs optional.

My suggestion was to:

* optional urls: min=1, max=1000, null=true (so `url=null`)
* required urls: min=1, max=1000

Your suggestion is:

* optional urls: min=0, max=1000  (so `url=""`)
* required urls: min=1, max=1000

You were rethinking this when you noticed the `status` field in the Zaak-resource, which can now be `null` if the zaak had no status.

**Comment by sergei-maertens** [comment]
@stevenbal so, summarized on what needs to be done:

* A subclass of `HyperlinkedRelatedField` should be used in the (HyperlinkedModel)Serializers: https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L1612. This would require all existing serializers to have to be updated, so I'm okay with monkey-patching `HyperlinkedModelSerializer.serializer_related_field` in the `apps.py` of vng-api-common.
* The `HyperlinkedRelatedField` subclass should accept the following kwargs:
    * `min_length`: default to 1
    * `max_length`: default to 1000
* the field can be `null` if it's not required - the meaning of `null` is "the related object does not exist". Clients passing an empty string for an optional field should pass `null` instead. This means that you should `kwargs.setdefault("allow_null", True)`, so that it can still be overridden on an individual basis if needed.
* the `min_length` and `max_length` need to be validated, of course
* a custom inspector for the custom field class can then extract the `min_length` and `max_length` and `allow_null`

File: vng_api_common/apps.py [hunk 1]
```diff
@@ -58,3 +59,9 @@ def register_serializer_field():
 
     if RelativeDeltaField is not None:
         mapping[RelativeDeltaField] = DurationField
+
+
+def set_custom_hyperlinkedmodelserializer_field():
```
Comment (sergei-maertens) [inline vng_api_common/apps.py:64]: please add a comment that this monkey-patches DRF to avoid having to change all the base classes in existing code

File: vng_api_common/inspectors/fields.py [hunk 2]
```diff
@@ -98,6 +98,31 @@ def field_to_swagger_object(
...
+    ):
+        SwaggerType, ChildSwaggerType = self._get_partial_types(
+            field, swagger_object_type, use_references, **kwargs
+        )
+
+        if (
+            isinstance(field, LengthHyperlinkedRelatedField)
+            and swagger_object_type == openapi.Schema
+        ):
+            max_length = getattr(field, "max_length", None)
```
Comment (sergei-maertens) [inline vng_api_common/inspectors/fields.py:113]: if it's this particular subclass, we _know_ that the attribute exists

```suggestion
            max_length = field.max_length
```

File: vng_api_common/inspectors/fields.py [hunk 3]
```diff
@@ -98,6 +98,31 @@ def field_to_swagger_object(
...
+            min_length = getattr(field, "min_length", None)
+            return SwaggerType(
```
Comment (sergei-maertens) [inline vng_api_common/inspectors/fields.py:115]: aren't we misisng the nullable support here? Or does the `FieldInspector` already do that?

File: vng_api_common/serializers.py [hunk 4]
```diff
@@ -264,3 +264,34 @@ def update(self, instance, validated_data):
                 continue
             setattr(instance, name, validated_data.pop(name))
         return super().update(instance, validated_data)
+
+
+class LengthHyperlinkedRelatedField(serializers.HyperlinkedRelatedField):
+    def __init__(self, *args, **kwargs):
+        self.min_length = kwargs.pop("min_length", None)
```
Comment (sergei-maertens) [inline vng_api_common/serializers.py:271]:
```suggestion
        self.min_length = kwargs.pop("min_length", 1)
```

File: vng_api_common/serializers.py [hunk 5]
```diff
@@ -264,3 +264,34 @@ def update(self, instance, validated_data):
...
+        self.max_length = kwargs.pop("max_length", None)
```
Comment (sergei-maertens) [inline vng_api_common/serializers.py:272]:
```suggestion
        self.max_length = kwargs.pop("max_length", 1000)
```

File: vng_api_common/serializers.py [hunk 6]
```diff
@@ -264,3 +264,34 @@ def update(self, instance, validated_data):
...
+        kwargs.setdefault("allow_null", True)
```
Comment (sergei-maertens) [inline vng_api_common/serializers.py:273]: this should _only_ be done if the field is not required

```suggestion
        kwargs.setdefault("allow_null", not kwargs.get("required", True))  # please check what the default value is in DRF
```

File: vng_api_common/serializers.py [hunk 7]
```diff
@@ -264,3 +264,34 @@ def update(self, instance, validated_data):
...
+
+        super().__init__(*args, **kwargs)
+
+    def to_internal_value(self, data):
+        if self.max_length and len(data) > self.max_length:
+            self.fail("max_length", max_length=self.max_length, length=len(data))
```
Comment (sergei-maertens) [inline vng_api_common/serializers.py:279]: where are these error messages defined?

File: vng_api_common/serializers.py [hunk 8]
```diff
@@ -264,3 +264,34 @@ def update(self, instance, validated_data):
...
+            # rewrite validation code to make it fit reference implementation
+            # if url is not valid
+            try:
+                URLValidator()(data)
+            except ValidationError as exc:
+                self.fail("bad-url", url=data, exc=exc)
+
+            # if the url is not bad -> then the problem is that it doesn't fit resource
+            self.fail("invalid-resource", exc=field_exc)
+        return value
```
Comment (sergei-maertens) [inline vng_api_common/serializers.py:297]: why is this needed?

File: vng_api_common/inspectors/fields.py (hunk 3 above)
Comment (stevenbal) [reply vng_api_common/inspectors/fields.py:115]: It does, I tested it with `allow_null` in the extra serializer kwargs, and it changes the api spec accordingly

File: vng_api_common/serializers.py (hunk 7 above)
Comment (stevenbal) [reply vng_api_common/serializers.py:279]: whoops, went a bit too hard on the copypaste

File: vng_api_common/serializers.py [hunk 9]
```diff
@@ -264,3 +265,28 @@ def update(self, instance, validated_data):
...
             setattr(instance, name, validated_data.pop(name))
         return super().update(instance, validated_data)
+
+
+class LengthHyperlinkedRelatedField(serializers.HyperlinkedRelatedField):
+    default_error_messages = {
+        "max_length": _("Ensure this field has no more than {max_length} characters."),
+        "min_length": _("Ensure this field has at least {min_length} characters."),
+        "bad-url": "The URL {url} could not be fetched. Exception: {exc}",
+        "invalid-resource": "Please provide a valid URL. Exception: {exc}",
```
Comment (sergei-maertens) [inline vng_api_common/serializers.py:275]: These aren't used, right?
